### PR TITLE
storage: avoid give_container in production of source data

### PR DESCRIPTION
**Pull request by guswynn**
Blocker on https://github.com/MaterializeInc/materialize/pull/26606

As @petrosagg said:
> This is an anti-pattern and must be fixed if it's the case. We used to do that in the pg source but I fixed it recently. I'm not aware of other cases. We should rely on it as an assumption for this PR

This removes our uses of `give_container` in the source data production side. It leaves it in upsert/decode/persist_sink/remap/source_render, which all appear to process data generally in the same batch as the input.

### Motivation

   * This PR refactors existing code.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage/src/source/source_reader_pipeline.rs [hunk 1]
```diff
@@ -345,14 +345,18 @@ where
...
             for statuses in statuses_by_idx.values_mut() {
                 if statuses.is_empty() {
                     continue;
                 }
 
-                health_output.give_container(&health_cap, statuses).await;
+                for u in statuses.drain(..) {
+                    health_output.give(&health_cap, u).await;
+                }
                 statuses.clear()
```
Comment (rjobanp) [inline src/storage/src/source/source_reader_pipeline.rs:360]: you don't need the `statuses.clear()` anymore considering the `drain(..)`

**Comment by guswynn** [comment]
@rjobanp can you take another look? I changed it so:
- we only avoid `give_container` on the source production side; operators in `source_reader_pipeline.rs` are mostly just forwarding things
- We use iterators to avoid manual container management in the mysql and `KEY VALUE` sources

**Comment by guswynn** [comment]
The agent lost error on the mysql resumption tests is spooky...I'm retrying, if it happens again ill look closer

**Comment by rjobanp** [comment]
New changes LGTM - also spooked that such a simple change is causing the lost agent on the mysql resumption tests..

**Comment by guswynn** [comment]
@petrosagg @rjobanp do yall mind taking another look at this?

File: src/storage/src/source/mysql/replication/events.rs [hunk 2]
```diff
@@ -281,18 +291,30 @@ pub(super) async fn handle_rows_event(
...
-    let gtid_cap = ctx.data_cap_set.delayed(new_gtid);
-    trace!(%id, "timely-{worker_id} sending container for {new_gtid:?} \
-                 with {} updates ({} additions, {} retractions) and {} \
-                 rewinds", container.len(), additions, retractions, rewind_count);
-    ctx.data_output
-        .give_container(&gtid_cap, &mut container)
-        .await;
+    // We want to emit data in individual pieces to allow timely to break large chunks of data into
+    // containers. Naively emitting data in the loop above interleaves rewinds and data emissions,
+    // which defeats a timely optimization that caches push buffers if the `.give()` time has not
```
Comment (petrosagg) [inline src/storage/src/source/mysql/replication/events.rs:300]: I think we should do it the other way around: give the normal data with `handle.give()` and buffer the rewind data which is rarer. Once all rewinds are done we can deallocate the buffer completely

Comment (guswynn) [reply src/storage/src/source/mysql/replication/events.rs:300]: ah true, that'll use less memory!

**Review by petrosagg — APPROVED** [review]

**Pull request merged**